File: frontend/apps/web/src/lib/api/policies.ts
/**
 * Policy API
 * ==========
 * API functions for expense policy management.
 */

import { api } from './client';

export interface Policy {
  id: string;
  company_id: string;
  name: string;
  amount_cap_usd: number;
  monthly_cap_usd?: number;
  allowed_categories: string[];
  vendor_whitelist: string[];
  vendor_blacklist: string[];
  max_days_old: number;
  custom_rules?: string;
  require_description: boolean;
  require_category: boolean;
  auto_approve_under: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreatePolicyRequest {
  company_id: string;
  name?: string;
  amount_cap_usd?: number;
  monthly_cap_usd?: number;
  allowed_categories?: string[];
  vendor_whitelist?: string[];
  vendor_blacklist?: string[];
  max_days_old?: number;
  custom_rules?: string;
  require_description?: boolean;
  require_category?: boolean;
  auto_approve_under?: number;
}

export interface UpdatePolicyRequest {
  name?: string;
  amount_cap_usd?: number;
  monthly_cap_usd?: number;
  allowed_categories?: string[];
  vendor_whitelist?: string[];
  vendor_blacklist?: string[];
  max_days_old?: number;
  custom_rules?: string;
  require_description?: boolean;
  require_category?: boolean;
  auto_approve_under?: number;
  is_active?: boolean;
}

export const policyApi = {
  /**
   * Create a new policy
   */
  async create(data: CreatePolicyRequest) {
    return api.post<Policy>('/policies', data);
  },

  /**
   * Get policy by ID
   */
  async getById(policyId: string) {
    return api.get<Policy>(`/policies/${policyId}`);
  },

  /**
   * Get active policy for a company
   */
  async getActiveForCompany(companyId: string) {
    return api.get<Policy>(`/policies/company/${companyId}/active`);
  },

  /**
   * List all policies for a company
   */
  async listByCompany(companyId: string) {
    return api.get<{ success: boolean; data: Policy[]; total: number }>(
      `/policies/company/${companyId}`
    );
  },

  /**
   * Update a policy
   */
  async update(policyId: string, data: UpdatePolicyRequest) {
    return api.patch<Policy>(`/policies/${policyId}`, data);
  },

  /**
   * Delete a policy
   */
  async delete(policyId: string) {
    return api.delete<{ success: boolean; message: string }>(`/policies/${policyId}`);
  },
};

File: backend/app/api/endpoints/policies.py
"""
Policy Endpoints
================
Expense policy management.
"""

from fastapi import APIRouter, HTTPException

from app.core.exceptions import AppException
from app.schemas.policy import (
    PolicyCreate,
    PolicyResponse,
    PolicyUpdate,
)
from app.services.policy import PolicyService

router = APIRouter()


@router.post(
    "",
    response_model=PolicyResponse,
    status_code=201,
    summary="Create a new policy",
)
async def create_policy(data: PolicyCreate):
    """
    Create a new expense policy for a company.

    The new policy will automatically become active,
    and any existing active policy will be deactivated.
    """
    try:
        service = PolicyService()
        return await service.create(data)
    except AppException as e:
        raise HTTPException(status_code=e.status_code, detail=e.to_dict())


@router.get(
    "/{policy_id}",
    response_model=PolicyResponse,
    summary="Get policy by ID",
)
async def get_policy(policy_id: str):
    """Get policy details by UUID."""
    try:
        service = PolicyService()
        return await service.get_by_id(policy_id)
    except AppException as e:
        raise HTTPException(status_code=e.status_code, detail=e.to_dict())


@router.get(
    "/company/{company_id}/active",
    response_model=PolicyResponse,
    summary="Get active policy",
    description="Get the currently active policy for a company.",
)
async def get_active_policy(company_id: str):
    """
    Get the active expense policy for a company.
    This is the policy that the Auditor Agent uses for validation.
    """
    try:
        service = PolicyService()
        return await service.get_active_for_company(company_id)
    except AppException as e:
        raise HTTPException(status_code=e.status_code, detail=e.to_dict())


@router.patch(
    "/{policy_id}",
    response_model=PolicyResponse,
    summary="Update policy",
)
async def update_policy(policy_id: str, data: PolicyUpdate):
    """
    Update policy details.

    If `is_active` is set to true, this policy will become active
    and any other active policy for the company will be deactivated.
    """
    try:
        service = PolicyService()
        return await service.update(policy_id, data)
    except AppException as e:
        raise HTTPException(status_code=e.status_code, detail=e.to_dict())


@router.get(
    "/company/{company_id}",
    summary="List policies by company",
)
async def list_by_company(company_id: str):
    """List all policies for a company (including inactive)."""
    try:
        service = PolicyService()
        policies = await service.list_by_company(company_id)
        return {
            "success": True,
            "data": policies,
            "total": len(policies),
        }
    except AppException as e:
        raise HTTPException(status_code=e.status_code, detail=e.to_dict())


@router.delete(
    "/{policy_id}",
    summary="Delete policy",
)
async def delete_policy(policy_id: str):
    """Delete a policy."""
    try:
        service = PolicyService()
        success = await service.delete(policy_id)
        return {"success": success, "message": "Policy deleted"}
    except AppException as e:
        raise HTTPException(status_code=e.status_code, detail=e.to_dict())
